File: divide-and-conquer/merge_sort.c
// merge sort
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

int input_data[100];

// merge two array into one
int *  merge(int *arr, int left, int middle, int right) {
	int *t = (int*)malloc(sizeof(int) * (right + 1 - left));
	int i = left;
	int j = middle + 1;
	int pos = 0;
	if ( middle + 1 - left > right - middle ) {
		while(i <= middle) {
			if(j<=right && arr[j] < arr[i]) {
				t[pos++] = arr[j++];
				continue;
			}
			t[pos++] = arr[i++];
		}
		while(j<=right) {
			t[pos++] = arr[j++];
		}
	}else {
		while(j <= right) {
			if(i<=middle && arr[i] < arr[j]) {
				t[pos++] = arr[i++];
				continue;
			}
			t[pos++] = arr[j++];
		}
		while(i<=middle) {
			t[pos++] = arr[i++];
		}
	}
	
	for(i=0, j=left; i<pos; i++,j++) {
		arr[j] = t[i];
	}
	return t;
}

// merge sort
void merge_sort(int *arr, int left, int right) {
	if ( left == right) {
		return;
	}

	int middle = left + (right - left) / 2;

	merge_sort(arr, left, middle);
	merge_sort(arr, middle+1, right);
	int *tmp = merge(arr, left, middle, right);
	free(tmp);
}

// show a array
void show(int *arr, int left, int right) {
	for(int i=left; i<=right; i++) {
		printf("%d ", arr[i]);
	}
	printf("\n");
}

int main() {
	
	srand((unsigned)time(NULL));
	for(int i=0; i<100; i++) {
		input_data[i] = rand() % 100;
	}
	show(input_data, 0, 99);
	merge_sort(input_data, 0, 99);
	show(input_data, 0, 99);
	return 0;
}

File: divide-and-conquer/merge_sort.py
#!usr/bin/env python
''' merge sort  '''

def merge(l, left, middle, right):
	''' merge 'two array' into 'one array' '''

	#print 'left is %d  middle is %d  right is %d' %(left, middle, right)

	t = []
	i = left 
	j = middle + 1 

	if middle - left + 1 > right - middle:
		while i <= middle:
			if j <= right and  l[j] < l[i]:
				t.append(l[j])
				j+=1
				continue
			t.append(l[i])
			i+=1
		while j <= right:
			t.append(l[j])
			j+=1
	else:
		while j <= right:
			if i <= middle and l[i] < l[j]:
				t.append(l[i])
				i+=1
				continue
			t.append(l[j])
			j+=1
		while i <= middle:
			t.append(l[i])
			i+=1

	#print ' t is %d numbers  l is %d numbers ' % (len(t), right + 1 - left)
	#show(t, msg = ' show list t')
	#show(l, msg = ' show list l')
	i = left
	for item in t:
		l[i] = item
		i+=1


def merge_sort(l, left, right):
	''' merge sort '''

	if left == right:
		return 
	middle = (left + right) / 2;
	merge_sort(l, left, middle);
	merge_sort(l, middle+1, right);
	merge(l, left, middle, right);

def show(l, msg = ''):
	''' show a array or a list in 0 ... lenght '''

	for item in l:
		print item,
	print


if __name__ == '__main__':
	import random
	l = random.sample(xrange(100), 100)
	show(l)
	merge_sort(l, 0, 99)
	show(l)

